icu4j/main/classes/core/src/com/ibm/icu/number/NumberFormatterImpl.java

```diff
@@ -412,6 +412,14 @@ private static MicroPropsGenerator macrosToMicroGenerator(MacroProps macros, Mic
                 MeasureUnit unit = macros.unit;
                 if (macros.perUnit != null) {
                     unit = unit.product(macros.perUnit.reciprocal());
+                    // This isn't strictly necessary, but was what we specced
+                    // out when perUnit became a backward-compatibility thing:
+                    // unit/perUnit use case is only valid if both units are
+                    // built-ins, or the product is a built-in.
+                    if (unit.getType() == null && (macros.unit.getType() == null || macros.perUnit.getType() == null)) {
+                        throw new UnsupportedOperationException(
```

Comment:
UnsupportedOperationException might not be the best exception here.
IllegalStateException is more for when data expectations are violated.
UnsupportedOperationException is more for "this class doesn't support a specific feature" (e.g. writing values to an immutable array).

Reply:
In NumberFormatterApiTest.java, the oldest reference I found to UnsupportedOperationException was this:

```
    static void assertUndefinedSkeleton(UnlocalizedNumberFormatter f) {
        try {
            String skeleton = f.toSkeleton();
            fail("Expected toSkeleton to fail, but it passed, producing: " + skeleton);
        } catch (UnsupportedOperationException expected) {}
    }
```
This is from three years ago (March 2018):
https://github.com/unicode-org/icu/commit/c2fa8cacad7fdf0c929b5686306c68cd88360837

I suspect consistency with this behaviour makes sense, and changing the behaviour would impact client code.

To illustrate by usage examples, these errors are the kind of thing that client code *does* have access to. These lines are fine - capitals for variables I'm not defining here:
```
nf = NumberFormatter.forSkeleton("unit/hectometer");
nf = NumberFormatter.forSkeleton("unit/hectometer-per-second");
nf = NumberFormatter.unit(HECTOMETER);
nf = NumberFormatter.unit(HECTOMETER_PER_SECOND);
```

These are unsupported, and throw an UnsupportedOperationException:
```
nf = NumberFormatter.forSkeleton("unit/hectometer per-unit/duration-second");
nf = NumberFormatter.unit(HECTOMETER).perUnit(SECOND);
```